packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,461 @@
+<?php
+
+/**
+ * The topological sorter class.
+ *
+ * We create an in-memory index that contains offsets and lengths of items in the WXR.
+ * The indexer will also topologically sort posts so that the order we iterate over posts
+ * ensures we always get parents before their children.
+ */
+class WP_Topological_Sorter {
+
+	/**
+	 * The base name of the table.
+	 */
+	const TABLE_NAME = 'data_liberation_index';
+
+	/**
+	 * The option name for the database version.
+	 */
+	const OPTION_NAME = 'data_liberation_db_version';
+
+	/**
+	 * The current database version, to be used with dbDelta.
+	 */
+	const DB_VERSION = 1;
+
+	// Element types.
+	const ELEMENT_TYPE_POST     = 1;
+	const ELEMENT_TYPE_CATEGORY = 2;
+
+	/**
+	 * Variable for keeping counts of orphaned posts/attachments, it'll also be assigned as temporarly post ID.
+	 * To prevent duplicate post ID, we'll use negative number.
+	 *
+	 * @var int
+	 */
+	protected $orphan_post_counter = 0;
+
+	/**
+	 * Store the ID of the post ID currently being processed.
+	 *
+	 * @var int
+	 */
+	protected $last_post_id = 0;
+
+	/**
+	 * Whether the sort has been done.
+	 *
+	 * @var bool
+	 */
+	protected $sorted = false;
+
+	/**
+	 * The current session ID.
+	 */
+	protected $current_session = null;
+
+	/**
+	 * The total number of categories.
+	 */
+	protected $total_categories = 0;
+
+	/**
+	 * The total number of posts.
+	 */
+	protected $total_posts = 0;
+
+	/**
+	 * The current item being processed.
+	 */
+	protected $current_item = 0;
+
+	public function __construct( $options = array() ) {
+		if ( array_key_exists( 'session_id', $options ) ) {
+			$this->current_session = $options['session_id'];
+		}
+	}
+
+	/**
+	 * Get the name of the table.
+	 *
+	 * @return string The name of the table.
+	 */
+	public static function get_table_name() {
+		global $wpdb;
+
+		// Default is wp_{TABLE_NAME}
+		return $wpdb->prefix . self::TABLE_NAME;
+	}
+
+	/**
+	 * Run by register_activation_hook.
+	 */
+	public static function activate() {
+		global $wpdb;
+
+		$table_name = self::get_table_name();
+
+		// Create the table if it doesn't exist.
+		// @TODO: remove this custom SQLite declaration after first phase of unit tests is done.
```

Comment:
I understand the CREATE TABLE statement doesn't work with the legacy SQLite integration we use now? cc @janjakes this could be useful for testing

Reply:
@adamziel @zaerl The MySQL query actually seems to work fine with the current translator.

Reply:
@JanJakes it is, thanks! Removed the special SQL.